// antd-mini/less/Form/FormInput/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var props_1 = require("./props");
var form_1 = require("../form");
var fmtEvent_1 = tslib_1.__importDefault(require("../../_util/fmtEvent"));
Component({
    props: props_1.FormInputDefaultProps,
    mixins: [(0, form_1.createForm)({
            methods: {
                setFormData: function (values) {
                    this.setData(tslib_1.__assign(tslib_1.__assign({}, this.data), { formData: tslib_1.__assign(tslib_1.__assign({}, this.data.formData), values) }));
                    this.input.update(this.data.formData.value);
                },
            },
        })],
    methods: {
        handleRef: function (input) {
            this.input = input;
        },
        onChange: function (value, e) {
            this.emit('onChange', value);
            if (this.props.onChange) {
                this.props.onChange(value, (0, fmtEvent_1.default)(this.props, e));
            }
        },
        onBlur: function (value, e) {
            if (this.props.onBlur) {
                this.props.onBlur(value, (0, fmtEvent_1.default)(this.props, e));
            }
        },
        onFocus: function (value, e) {
            if (this.props.onChange) {
                this.props.onFocus(value, (0, fmtEvent_1.default)(this.props, e));
            }
        },
        onConfirm: function (value, e) {
            if (this.props.onConfirm) {
                this.props.onConfirm(value, (0, fmtEvent_1.default)(this.props, e));
            }
        },
        onClear: function (value, e) {
            this.emit('onChange', '');
            if (this.props.onChange) {
                this.props.onChange(value, (0, fmtEvent_1.default)(this.props, e));
            }
        },
    }
});
