// antd-mini/less/Button/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var props_1 = require("./props");
var fmtEvent_1 = tslib_1.__importDefault(require("../_util/fmtEvent"));
Component({
    props: props_1.ButtonDefaultProps,
    methods: {
        onTap: function (e) {
            var _a = this.props, onTap = _a.onTap, disabled = _a.disabled, loading = _a.loading;
            if (onTap && !disabled && !loading) {
                var event = (0, fmtEvent_1.default)(this.props, e);
                return onTap(event);
            }
        },
        catchTap: function (e) {
            var _a = this.props, catchTap = _a.catchTap, disabled = _a.disabled, loading = _a.loading;
            if (catchTap && !disabled && !loading) {
                var event = (0, fmtEvent_1.default)(this.props, e);
                return catchTap(event);
            }
        },
        onGetAuthorize: function (e) {
            if (this.props.onGetAuthorize) {
                this.props.onGetAuthorize((0, fmtEvent_1.default)(this.props, e));
            }
        },
        onFollowLifestyle: function (e) {
            if (this.props.onFollowLifestyle) {
                this.props.onFollowLifestyle((0, fmtEvent_1.default)(this.props, e));
            }
        },
        onError: function (e) {
            if (this.props.onError) {
                this.props.onError((0, fmtEvent_1.default)(this.props, e));
            }
        },
        onGetUserInfo: function (e) {
            if (this.props.onGetUserInfo) {
                this.props.onGetUserInfo((0, fmtEvent_1.default)(this.props, e));
            }
        },
        onGetPhoneNumber: function (e) {
            if (this.props.onGetPhoneNumber) {
                this.props.onGetPhoneNumber((0, fmtEvent_1.default)(this.props, e));
            }
        },
    },
});
