// antd-mini/less/Form/FormPicker/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var props_1 = require("./props");
var createComponent_1 = tslib_1.__importDefault(require("../createComponent"));
var fmtEvent_1 = tslib_1.__importDefault(require("../../_util/fmtEvent"));
(0, createComponent_1.default)({
    props: props_1.FormPickerDefaultProps,
    methods: {
        onOk: function (value, column, e) {
            this.emit('onChange', value);
            if (this.props.onOk) {
                this.props.onOk(value, column, (0, fmtEvent_1.default)(this.props, e));
            }
        },
        onChange: function (value, column, e) {
            if (this.props.onChange) {
                this.props.onChange(value, column, (0, fmtEvent_1.default)(this.props, e));
            }
        },
        onVisibleChange: function (visible, e) {
            if (this.props.onVisibleChange) {
                this.props.onVisibleChange(visible, (0, fmtEvent_1.default)(this.props, e));
            }
        },
        onFormat: function (value, column) {
            if (this.props.onFormat) {
                return this.props.onFormat(value, column);
            }
        },
        onDismissPicker: function (e) {
            if (this.props.onCancel) {
                this.props.onCancel((0, fmtEvent_1.default)(this.props, e));
            }
        },
    },
});
