// antd-mini/less/Modal/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var props_1 = require("./props");
Component({
    props: props_1.ModalDefaultProps,
    methods: {
        onClose: function () {
            var onClose = this.props.onClose;
            if (onClose) {
                onClose();
            }
        },
        onMaskClose: function () {
            var _a = this.props, onClose = _a.onClose, maskClosable = _a.maskClosable;
            if (maskClosable && onClose) {
                onClose();
            }
        },
        onPrimaryButtonTap: function () {
            var onPrimaryButtonTap = this.props.onPrimaryButtonTap;
            if (onPrimaryButtonTap) {
                onPrimaryButtonTap();
            }
        },
        onSecondaryButtonTap: function () {
            var onSecondaryButtonTap = this.props.onSecondaryButtonTap;
            if (onSecondaryButtonTap) {
                onSecondaryButtonTap();
            }
        },
        onCancelButtonTap: function () {
            var onCancelButtonTap = this.props.onCancelButtonTap;
            if (onCancelButtonTap) {
                onCancelButtonTap();
            }
        },
    },
});
