// antd-mini/es/Checklist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var props_1 = require("./props");
var value_1 = tslib_1.__importDefault(require("../mixins/value"));
var fmtEvent_1 = tslib_1.__importDefault(require("../_util/fmtEvent"));
Component({
    props: props_1.ChecklistDefaultProps,
    mixins: [
        (0, value_1.default)({
            transformValue: function (val) {
                var value = val || [];
                return {
                    needUpdate: true,
                    value: value,
                };
            },
        }),
    ],
    methods: {
        onChange: function (item) {
            var _a = this.props, multiple = _a.multiple, options = _a.options, onChange = _a.onChange;
            var value = item.value;
            if (multiple) {
                var currentValue_1 = this.getValue();
                if (currentValue_1.indexOf(value) > -1) {
                    currentValue_1 = currentValue_1.filter(function (v) { return v !== value; });
                }
                else {
                    currentValue_1 = tslib_1.__spreadArray(tslib_1.__spreadArray([], currentValue_1, true), [value], false);
                }
                if (!this.isControlled()) {
                    this.update(currentValue_1);
                }
                if (onChange) {
                    onChange(currentValue_1, options.filter(function (v) { return currentValue_1.indexOf(v.value) > -1; }), (0, fmtEvent_1.default)(this.props));
                }
            }
            else {
                if (!this.isControlled()) {
                    this.update(value);
                }
                if (onChange) {
                    onChange(value, options.find(function (v) { return v.value === value; }), (0, fmtEvent_1.default)(this.props));
                }
            }
        },
    },
});
