// antd-mini/es/Form/form.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createForm = exports.Form = exports.Values = exports.Value = void 0;
var tslib_1 = require("tslib");
var async_validator_1 = tslib_1.__importStar(require("async-validator"));
Object.defineProperty(exports, "Value", { enumerable: true, get: function () { return async_validator_1.Value; } });
Object.defineProperty(exports, "Values", { enumerable: true, get: function () { return async_validator_1.Values; } });
var EventEmitter = /** @class */ (function () {
    function EventEmitter() {
        this.listenders = {};
    }
    EventEmitter.prototype.on = function (event, listener) {
        this.listenders[event] = this.listenders[event] || [];
        this.listenders[event].push(listener);
        return this;
    };
    EventEmitter.prototype.emit = function (event) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        var arr = this.listenders[event];
        if (arr) {
            arr.forEach(function (listener) { return listener.apply(void 0, args); });
        }
    };
    return EventEmitter;
}());
var Field = /** @class */ (function (_super) {
    tslib_1.__extends(Field, _super);
    /**
     * Field构建
     * @param ref field ref对象
     * @param initialValue 初始值
     */
    function Field(ref, name, initialValues, rules, validateMessages, required, message, validateTrigger) {
        var _this = _super.call(this) || this;
        _this.ref = ref;
        _this.create(name, initialValues[name], rules[name], validateMessages, required, message, validateTrigger);
        _this.ref.on(function (trigger, value) {
            if (trigger === 'onChange') {
                _this.setValue(value);
                _this.touched = true;
                _this.emit('valueChange', value);
            }
            else if (trigger === 'didUnmount') {
                _this.emit('didUnmount');
            }
            else if (trigger === 'deriveDataFromProps') {
                var props = _this.ref.getProps();
                if (value.name && value.name !== props.name || value.required !== props.required || value.message !== props.message || value.validateTrigger !== props.validateTrigger) {
                    _this.create(value.name, initialValues[value.name], rules[value.name], validateMessages, value.required, value.message, value.validateTrigger);
                }
                if (value.name !== props.name) {
                    _this.emit('replaceName', value.name);
                }
            }
            _this.validateTrigger.forEach(function (item) {
                if (item === trigger) {
                    _this.validate();
                }
            });
        });
        return _this;
    }
    Field.prototype.create = function (name, initialValue, rule, validateMessages, required, message, validateTrigger) {
        this.name = name;
        this.required = this.transformValidatorRules(name, rule, required, message, validateMessages);
        this.reset(initialValue);
        var validateTriggerList = validateTrigger || 'onChange';
        if (!Array.isArray(validateTriggerList)) {
            validateTriggerList = [validateTriggerList];
        }
        this.validateTrigger = validateTriggerList;
    };
    /**
     *
     * @param rule 修改 Validator
     * @param name
     * @param required
     * @param message
     * @param validateMessages
     * @returns
     */
    Field.prototype.transformValidatorRules = function (name, rule, required, message, validateMessages) {
        var _a, _b;
        var requiredRule = false;
        var validator;
        if (rule) {
            var ruleList = Array.isArray(rule) ? rule : [rule];
            var result = ruleList.find(function (item) { return item.required; });
            if (result) {
                if (message) {
                    result.message = message;
                }
                requiredRule = true;
            }
            else if (required) {
                ruleList.push({
                    required: required,
                    message: message,
                });
                requiredRule = true;
            }
            validator = new async_validator_1.default((_a = {},
                _a[name] = ruleList,
                _a));
        }
        else if (required) {
            validator = new async_validator_1.default((_b = {},
                _b[name] = {
                    required: required,
                    message: message,
                },
                _b));
            requiredRule = true;
        }
        if (validator) {
            validator.messages(validateMessages);
        }
        this.validator = validator;
        return requiredRule;
    };
    /**
     * 设置 Field 值
     * @param value Field 值
     */
    Field.prototype.setValue = function (value) {
        this.ref.setFormData({
            value: value,
        });
    };
    /**
     * 得到 Field 值
     */
    Field.prototype.getValue = function () {
        var value = this.ref.getFormData().value;
        return value;
    };
    /**
     * 设置 Field 校验器状态
     * @param validatorStatue
     */
    Field.prototype.setValidatorStatus = function (validatorStatue) {
        this.ref.setFormData(validatorStatue);
    };
    /**
     * 得到 Field 校验器状态
     * @returns
     */
    Field.prototype.getValidatorStatus = function () {
        var _a = this.ref.getFormData(), status = _a.status, errors = _a.errors;
        return {
            status: status,
            errors: errors,
        };
    };
    /**
     * 校验 Field
     */
    Field.prototype.validate = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var validatorStatusSuccess, value, validator, err_1, errors, validatorStatus;
            var _a;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        validatorStatusSuccess = {
                            status: 'success',
                            errors: [],
                        };
                        value = this.getValue();
                        if (!this.validator) {
                            this.setValidatorStatus(validatorStatusSuccess);
                            return [2 /*return*/, {
                                    validatorStatus: validatorStatusSuccess,
                                    value: value,
                                }];
                        }
                        validator = this.validator;
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 3, , 4]);
                        this.setValidatorStatus({
                            status: 'validating',
                            errors: []
                        });
                        return [4 /*yield*/, this.validator.validate((_a = {},
                                _a[this.name] = value,
                                _a))];
                    case 2:
                        _b.sent();
                        if (validator !== this.validator) {
                            return [2 /*return*/];
                        }
                        this.setValidatorStatus(validatorStatusSuccess);
                        return [2 /*return*/, {
                                validatorStatus: validatorStatusSuccess,
                                value: value,
                            }];
                    case 3:
                        err_1 = _b.sent();
                        if (validator !== this.validator) {
                            return [2 /*return*/];
                        }
                        errors = err_1.errors;
                        validatorStatus = {
                            status: 'error',
                            errors: errors.map(function (_a) {
                                var _b = _a.message, message = _b === void 0 ? '' : _b;
                                return message;
                            }),
                        };
                        this.setValidatorStatus(validatorStatus);
                        return [2 /*return*/, {
                                validatorStatus: validatorStatus,
                                value: value,
                            }];
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * 重置 Field
     * @param initialValue
     */
    Field.prototype.reset = function (initialValue) {
        this.touched = false;
        this.ref.setFormData({
            value: initialValue,
            required: this.required,
            status: 'default',
            errors: [],
        });
    };
    /**
     * Field 是否被操作
     */
    Field.prototype.isTouched = function () {
        return this.touched;
    };
    return Field;
}(EventEmitter));
var Form = /** @class */ (function () {
    /**
     * Form构建
     * @param formConfig 表单配置项
     */
    function Form(formConfig) {
        if (formConfig === void 0) { formConfig = {}; }
        /**
         * 表单ref组件对象
         */
        this.fields = {};
        /**
         * 表单字段 change侦听
         */
        this.changeListeners = {};
        /**
         * 依赖表
         */
        this.dependenciesMap = {};
        this.setInitialValues(formConfig.initialValues || {});
        this.setRules(formConfig.rules || {});
        this.validateMessages = formConfig.validateMessages;
    }
    /**
     * 用户传入的rules转换成async-validator rules
     * @param rules 校验规则
     */
    Form.prototype.transformRules = function (rules) {
        var _this = this;
        var result = {};
        Object.keys(rules).forEach(function (name) {
            var rule = rules[name];
            var list = result[name] = [];
            var arr = Array.isArray(rule) ? rule : [rule];
            arr.forEach(function (item) {
                if (typeof item === 'function') {
                    list.push(item(_this).validator);
                }
                else {
                    list.push(tslib_1.__assign({}, item));
                }
            });
        });
        return result;
    };
    /**
     * 遍历表单field对象
     * @param callback
     */
    Form.prototype.eachField = function (callback) {
        var fields = this.fields;
        Object.keys(fields).forEach(function (name) {
            var field = fields[name];
            callback(field, name);
        });
    };
    /**
     * 设置 rules
     * @param rules
     */
    Form.prototype.setRules = function (rules) {
        this.rules = this.transformRules(rules);
    };
    /**
     * 添加表单对象
     * @param ref 表单ref对象
     */
    Form.prototype.addItem = function (ref, customName) {
        var _this = this;
        var props = ref.getProps();
        var name = customName || props.name;
        if (!name) {
            ref.on(function (trigger, value) {
                if (trigger === 'deriveDataFromProps') {
                    if (value.name) {
                        _this.addItem(ref, value.name);
                    }
                }
            });
            return;
        }
        if (this.fields[name]) {
            throw new Error("Form \"addItem\" same name: \"".concat(name, "\""));
        }
        var field = new Field(ref, name, this.initialValues, this.rules, this.validateMessages, props.required, props.message, props.validateTrigger);
        if (props.dependencies) {
            props.dependencies.forEach(function (item) {
                _this.dependenciesMap[item] = _this.dependenciesMap[item] || [];
                if (_this.dependenciesMap[item].indexOf(name) < 0) {
                    _this.dependenciesMap[item].push(name);
                }
            });
        }
        field.on('valueChange', function (value) {
            if (name) {
                var arr = _this.dependenciesMap[name];
                if (arr) {
                    arr.forEach(function (item) {
                        if (_this.fields[item]) {
                            _this.fields[item].validate();
                        }
                    });
                }
                if (_this.changeListeners[name]) {
                    _this.changeListeners[name].forEach(function (item) { return item(value); });
                }
            }
        }).on('didUnmount', function () {
            delete _this.fields[name];
        }).on('replaceName', function (newName) {
            if (!newName) {
                delete _this.fields[name];
                return;
            }
            if (_this.fields[newName]) {
                throw new Error("Form \"addItem\" same name: \"".concat(newName, "\""));
            }
            _this.fields[newName] = field;
            delete _this.fields[name];
            name = newName;
        });
        if (name) {
            this.fields[name] = field;
        }
    };
    /**
     * 设置表单值
     * @param name 表单名称
     * @param value 表单初始值
     */
    Form.prototype.setFieldValue = function (name, value) {
        var field = this.fields[name];
        if (field) {
            field.setValue(value);
            field.validate();
        }
    };
    /**
     * 设置 initialValues，这个操作不会对页面进行修改，要是需要重置表单可跟上 reset 方法；
     * 这样是对于表单已经在编辑，但是需要重新initialValues的场景
     *
     * eg:
     *    this.setInitialValues(initialValues);
     *    this.reset();
     *
     * @param initialValues
     */
    Form.prototype.setInitialValues = function (initialValues) {
        this.initialValues = initialValues;
    };
    /**
     * 获取对应字段名的值
     * @param name
     * @returns
     */
    Form.prototype.getFieldValue = function (name) {
        var field = this.fields[name];
        if (!field) {
            return;
        }
        return field.getValue();
    };
    /**
     * 获取对应字段名的校验器状态
     * @param name
     * @returns
     */
    Form.prototype.getFieldValidatorStatus = function (name) {
        var field = this.fields[name];
        if (!field) {
            return;
        }
        return field.getValidatorStatus();
    };
    /**
     * 检查对应字段是否被用户操作过
     * @param name 字段名称
     * @returns
     */
    Form.prototype.isFieldTouched = function (name) {
        var field = this.fields[name];
        if (!field) {
            return false;
        }
        return field.isTouched();
    };
    /**
     * 表单字段值更新时触发回调方法
     * @param name 表单字段名称
     * @param callback 回调方法
     */
    Form.prototype.onValueChange = function (name, callback) {
        this.changeListeners[name] = this.changeListeners[name] || [];
        this.changeListeners[name].push(callback);
    };
    /**
     * 表单提交
     */
    Form.prototype.submit = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var values, arr, result, errorFields;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        values = {};
                        arr = [];
                        this.eachField(function (field, name) {
                            arr.push((function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                                var _a;
                                return tslib_1.__generator(this, function (_b) {
                                    switch (_b.label) {
                                        case 0:
                                            _a = [{}];
                                            return [4 /*yield*/, field.validate()];
                                        case 1: return [2 /*return*/, tslib_1.__assign.apply(void 0, [tslib_1.__assign.apply(void 0, _a.concat([_b.sent()])), { name: name }])];
                                    }
                                });
                            }); })());
                        });
                        return [4 /*yield*/, Promise.all(arr)];
                    case 1:
                        result = _a.sent();
                        errorFields = [];
                        result.forEach(function (obj) {
                            if (!obj) {
                                return;
                            }
                            var name = obj.name, value = obj.value, validatorStatus = obj.validatorStatus;
                            if (validatorStatus.status === 'error') {
                                errorFields.push({
                                    name: name,
                                    errors: validatorStatus.errors,
                                });
                            }
                            values[name] = value;
                        });
                        if (errorFields.length > 0) {
                            throw {
                                values: values,
                                errorFields: errorFields,
                            };
                        }
                        return [2 /*return*/, values];
                }
            });
        });
    };
    /**
     * 表单重置
     */
    Form.prototype.reset = function () {
        var _this = this;
        this.eachField(function (field, name) {
            var initialValue = _this.initialValues[name];
            field.reset(initialValue);
        });
    };
    return Form;
}());
exports.Form = Form;
function createForm(_a) {
    var _b = _a === void 0 ? {} : _a, _c = _b.methods, methods = _c === void 0 ? {} : _c;
    return {
        data: {
            formData: {
                value: undefined,
                status: 'default',
                errors: [],
            },
        },
        didUnmount: function () {
            this.emit('didUnmount');
        },
        deriveDataFromProps: function (nextProps) {
            this.emit('deriveDataFromProps', nextProps);
        },
        methods: tslib_1.__assign({ emit: function (trigger, value) { }, setFormData: function (values) {
                this.setData(tslib_1.__assign(tslib_1.__assign({}, this.data), { formData: tslib_1.__assign(tslib_1.__assign({}, this.data.formData), values) }));
            }, getFormData: function () {
                return this.data.formData;
            }, on: function (callback) {
                this.emit = callback;
            }, getProps: function () {
                return this.props;
            } }, methods)
    };
}
exports.createForm = createForm;
