// antd-mini/es/Form/FormImageUpload/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var props_1 = require("./props");
var createComponent_1 = tslib_1.__importDefault(require("../createComponent"));
(0, createComponent_1.default)({
    props: props_1.FormImageUploadDefaultProps,
    methods: {
        onChange: function (value) {
            this.emit('onChange', value);
            if (this.props.onChange) {
                this.props.onChange(value);
            }
        },
        onUpload: function (localFile) {
            if (!this.props.onUpload) {
                throw new Error('need props onUpload');
            }
            return this.props.onUpload(localFile);
        },
        onRemove: function (file) {
            if (this.props.onRemove) {
                this.props.onRemove(file);
            }
        },
        onPreview: function (file) {
            if (this.props.onPreview) {
                this.props.onPreview(file);
            }
        },
        onBeforeUpload: function (localFileList) {
            if (this.props.onBeforeUpload) {
                return this.props.onBeforeUpload(localFileList);
            }
        },
        onChooseImageError: function (err) {
            if (this.props.onChooseImageError) {
                this.props.onChooseImageError(err);
            }
        },
    }
});
