// antd-mini/es/Checkbox/CheckboxGroup/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var props_1 = require("./props");
var fmtEvent_1 = tslib_1.__importDefault(require("../../_util/fmtEvent"));
var value_1 = tslib_1.__importDefault(require("../../mixins/value"));
Component({
    props: props_1.CheckboxGroupDefaultProps,
    mixins: [
        (0, value_1.default)({
            transformValue: function (val) {
                var value = val || [];
                return {
                    needUpdate: true,
                    value: value,
                };
            },
        }),
    ],
    methods: {
        onChange: function (_, e) {
            if (this.props.disabled) {
                return;
            }
            var currentValue = this.getValue();
            var index = e.currentTarget.dataset.index;
            var value = this.props.options[index].value;
            if (currentValue.indexOf(value) > -1) {
                currentValue = currentValue.filter(function (v) { return v !== value; });
            }
            else {
                currentValue = tslib_1.__spreadArray(tslib_1.__spreadArray([], currentValue, true), [value], false);
            }
            if (!this.isControlled()) {
                this.update(currentValue);
            }
            if (this.props.onChange) {
                this.props.onChange(currentValue, (0, fmtEvent_1.default)(this.props, e));
            }
        },
    },
});
