// antd-mini/less/Form/FormStepper/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var props_1 = require("./props");
var createComponent_1 = tslib_1.__importDefault(require("../createComponent"));
var fmtEvent_1 = tslib_1.__importDefault(require("../../_util/fmtEvent"));
(0, createComponent_1.default)({
    props: props_1.FormStepperDefaultProps,
    methods: {
        onChange: function (value, e) {
            this.emit('onChange', value);
            if (this.props.onChange) {
                this.props.onChange(value, (0, fmtEvent_1.default)(this.props, e));
            }
        },
        onBlur: function (e) {
            if (this.props.onBlur) {
                this.props.onBlur((0, fmtEvent_1.default)(this.props, e));
            }
        },
        onFocus: function (e) {
            if (this.props.onChange) {
                this.props.onFocus((0, fmtEvent_1.default)(this.props, e));
            }
        },
        onConfirm: function (e) {
            if (this.props.onConfirm) {
                this.props.onConfirm((0, fmtEvent_1.default)(this.props, e));
            }
        },
    }
});
