// antd-mini/es/Form/FormDatePicker/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var props_1 = require("./props");
var createComponent_1 = tslib_1.__importDefault(require("../createComponent"));
var fmtEvent_1 = tslib_1.__importDefault(require("../../_util/fmtEvent"));
(0, createComponent_1.default)({
    props: props_1.FormDatePickerDefaultProps,
    methods: {
        onOk: function (date, dateStr, e) {
            this.emit('onChange', date);
            if (this.props.onOk) {
                this.props.onOk(date, dateStr, (0, fmtEvent_1.default)(this.props, e));
            }
        },
        onPickerChange: function (date, dateStr, e) {
            if (this.props.onPickerChange) {
                this.props.onPickerChange(date, dateStr, (0, fmtEvent_1.default)(this.props, e));
            }
        },
        onVisibleChange: function (visible, e) {
            if (this.props.onVisibleChange) {
                this.props.onVisibleChange(visible, (0, fmtEvent_1.default)(this.props, e));
            }
        },
        onDismissPicker: function (e) {
            if (this.props.onCancel) {
                this.props.onDismissPicker((0, fmtEvent_1.default)(this.props, e));
            }
        },
        onFormat: function (date, dateStr) {
            if (this.props.onFormat) {
                return this.props.onFormat(date, dateStr);
            }
        },
    },
});
